### PR TITLE
adam support multithread

### PR DESCRIPTION
**Pull request by jacquesqiao**
test result:
200M parameter + sparse update 
4 thread speed up 4 times.

### CODE REVIEW

File: paddle/fluid/operators/optimizers/adam_op.h [hunk 1]
```diff
@@ -372,6 +373,28 @@ class AdamOpKernel : public framework::OpKernel<T> {
             functor.adam_update(i, grad_data[row_index * row_numel + offset]);
           }
         }
+      } else if (FLAGS_inner_op_parallelism > 1 &&
```
Comment (typhoonzero) [inline paddle/fluid/operators/optimizers/adam_op.h]: seems indent is not right for the if block

File: paddle/fluid/operators/optimizers/adam_op.h [hunk 2]
```diff
@@ -372,6 +373,28 @@ class AdamOpKernel : public framework::OpKernel<T> {
...
+                 FLAGS_min_param_size_to_use_multithread > 0 &&
+                 param.numel() > FLAGS_min_param_size_to_use_multithread) {
+        VLOG(3) << "use multi thread, inner_op_parallelism="
+                << FLAGS_inner_op_parallelism
+                << " min_param_size_to_use_multithread="
+                << FLAGS_min_param_size_to_use_multithread;
+        std::vector<std::future<void>> fs;
```
Comment (typhoonzero) [inline paddle/fluid/operators/optimizers/adam_op.h]: fs => futures maybe better?

File: paddle/fluid/operators/optimizers/adam_op.h [hunk 3]
```diff
@@ -372,6 +373,28 @@ class AdamOpKernel : public framework::OpKernel<T> {
             functor.adam_update(i, grad_data[row_index * row_numel + offset]);
           }
         }
+      } else if (FLAGS_inner_op_parallelism > 1 &&
+                 FLAGS_min_param_size_to_use_multithread > 0 &&
+                 param.numel() > FLAGS_min_param_size_to_use_multithread) {
```
Comment (typhoonzero) [inline paddle/fluid/operators/optimizers/adam_op.h]: `numel()` is slow, cache it to a var should be better

**Review by velconia — DISMISSED** [review]
LGTM

File: paddle/fluid/operators/optimizers/adam_op.h [hunk 4]
```diff
@@ -476,6 +479,62 @@ class AdamOpKernel : public framework::OpKernel<T> {
...
           }
+        } else if (FLAGS_inner_op_parallelism > 1 &&
+                   min_row_size_to_use_multithread > 0 &&
+                   param.dims()[0] > min_row_size_to_use_multithread) {
+          VLOG(3) << "use multi thread, inner_op_parallelism="
+                  << FLAGS_inner_op_parallelism
+                  << " min_row_size_to_use_multithread="
+                  << min_row_size_to_use_multithread;
+          if (FLAGS_inner_op_parallelism > 10) {
+            LOG(WARNING) << "FLAGS_inner_op_parallelism "
```
Comment (panyx0718) [inline paddle/fluid/operators/optimizers/adam_op.h]: Can there be a lot of warning?

File: paddle/fluid/operators/optimizers/adam_op.h [hunk 5]
```diff
@@ -476,6 +479,62 @@ class AdamOpKernel : public framework::OpKernel<T> {
...
+                         << FLAGS_inner_op_parallelism << " is two large!";
+          }
+          auto& grad_rows = grad_merge.rows();
+          std::unordered_map<size_t, int> row_id_to_grad_row_offset;
+          size_t param_row_count = param.numel() / row_numel;
+          if (param_row_count < 1000) {
+            LOG(WARNING) << "param_row_count should be larger then 1000 to use "
```
Comment (panyx0718) [inline paddle/fluid/operators/optimizers/adam_op.h]: same above

File: python/paddle/fluid/tests/unittests/test_adam_op.py [hunk 6]
```diff
@@ -253,15 +253,20 @@ def setup(self, scope, place, lazy_mode):
         row_numel = 12
         self.row_numel = row_numel
         self.dense_inputs = {
-            "Param": np.full((height, row_numel), 5.0).astype("float32"),
-            "Moment1": np.full((height, row_numel), 5.0).astype("float32"),
-            "Moment2": np.full((height, row_numel), 5.0).astype("float32"),
-            'Beta1Pow': np.array([beta1**10]).astype("float32"),
-            'Beta2Pow': np.array([beta2**10]).astype("float32"),
+            "Param": np.full((height, row_numel), 1.0).astype("float32"),
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/test_adam_op.py]: why change this? Is the multi-threading tested?

File: paddle/fluid/API.spec [hunk 7]
```diff
@@ -410,7 +410,7 @@ paddle.fluid.optimizer.MomentumOptimizer.__init__ ArgSpec(args=['self', 'learnin
 paddle.fluid.optimizer.MomentumOptimizer.minimize ArgSpec(args=['self', 'loss', 'startup_program', 'parameter_list', 'no_grad_set'], varargs=None, keywords=None, defaults=(None, None, None))
 paddle.fluid.optimizer.AdagradOptimizer.__init__ ArgSpec(args=['self', 'learning_rate', 'epsilon', 'regularization', 'name'], varargs=None, keywords=None, defaults=(1e-06, None, None))
 paddle.fluid.optimizer.AdagradOptimizer.minimize ArgSpec(args=['self', 'loss', 'startup_program', 'parameter_list', 'no_grad_set'], varargs=None, keywords=None, defaults=(None, None, None))
-paddle.fluid.optimizer.AdamOptimizer.__init__ ArgSpec(args=['self', 'learning_rate', 'beta1', 'beta2', 'epsilon', 'regularization', 'name', 'lazy_mode'], varargs=None, keywords=None, defaults=(0.001, 0.9, 0.999, 1e-08, None, None, False))
+paddle.fluid.optimizer.AdamOptimizer.__init__ ArgSpec(args=['self', 'learning_rate', 'beta1', 'beta2', 'epsilon', 'regularization', 'name', 'lazy_mode', 'min_row_size_to_use_multithread'], varargs=None, keywords=None, defaults=(0.001, 0.9, 0.999, 1e-08, None, None, False, 0))
```
Comment (panyx0718) [inline paddle/fluid/API.spec]: Can this be decided automatically?

File: paddle/fluid/operators/optimizers/adam_op.h [hunk 8]
```diff
@@ -476,6 +479,62 @@ class AdamOpKernel : public framework::OpKernel<T> {
...
+            int64_t start = i * line_in_each_thread;
+            int64_t end = (i + 1) * line_in_each_thread;
+            if (start >= param_row_count) {
+              break;
+            }
+            if (end > param_row_count) {
+              end = param_row_count;
+            }
+            fs.push_back(
+                framework::Async([&functor, &row_id_to_grad_row_offset,
```
Comment (velconia) [inline paddle/fluid/operators/optimizers/adam_op.h:524]: I guess this should be tested with ```FastThreadedExecutor``` together to validate there is no thread problem.

File: paddle/fluid/API.spec (hunk 7 above)
Comment (velconia) [reply paddle/fluid/API.spec]: Maybe removed in python and keep in C++?

File: python/paddle/fluid/tests/unittests/test_adam_op.py (hunk 6 above)
Comment (jacquesqiao) [reply python/paddle/fluid/tests/unittests/test_adam_op.py]: test by https://github.com/PaddlePaddle/Paddle/pull/14890/files#diff-68cdbd1cba20d9e8a5ddfcdd1ffa88f6R90

File: paddle/fluid/API.spec (hunk 7 above)
Comment (jacquesqiao) [reply paddle/fluid/API.spec]: done, add a default value to op attribute

File: python/paddle/fluid/tests/unittests/test_adam_op.py (hunk 6 above)
Comment (jacquesqiao) [reply python/paddle/fluid/tests/unittests/test_adam_op.py]: fix done

File: paddle/fluid/operators/optimizers/adam_op.h (hunk 3 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/optimizers/adam_op.h]: removed

**Comment by luotao1** [comment]
Could you fix the warning in `adam_op`?
```
[04:00:48][Step 1/1] In file included from /paddle/paddle/fluid/operators/optimizers/adam_op.cc:15:0:
[04:00:48][Step 1/1] /paddle/paddle/fluid/operators/optimizers/adam_op.h: In instantiation of 'void paddle::operators::SparseAdamFunctor<T, paddle::operators::CPUAdam>::operator()(size_t) const [with T = float; size_t = long unsigned int]':
[04:00:48][Step 1/1] /paddle/paddle/fluid/operators/optimizers/adam_op.h:487:18:   required from 'void paddle::operators::AdamOpKernel<DeviceContext, T>::Compute(const paddle::framework::ExecutionContext&) const [with DeviceContext = paddle::platform::CPUDeviceContext; T = float]'
[04:00:48][Step 1/1] /paddle/paddle/fluid/framework/op_registry.h:107:11:   required from 'paddle::framework::OpKernelRegistrarFunctor<PlaceType, false, I, KernelTypes ...>::operator()(const char*, const char*, int) const::<lambda(const paddle::framework::ExecutionContext&)> [with PlaceType = paddle::platform::CPUPlace; long unsigned int I = 0ul; KernelTypes = {paddle::operators::AdamOpKernel<paddle::platform::CPUDeviceContext, float>, paddle::operators::AdamOpKernel<paddle::platform::CPUDeviceContext, double>}]'
[04:00:48][Step 1/1] /paddle/paddle/fluid/framework/op_registry.h:106:10:   required from 'struct paddle::framework::OpKernelRegistrarFunctor<PlaceType, false, I, KernelTypes ...>::operator()(const char*, const char*, int) const [with PlaceType = paddle::platform::CPUPlace; long unsigned int I = 0ul; KernelTypes = {paddle::operators::AdamOpKernel<paddle::platform::CPUDeviceContext, float>, paddle::operators::AdamOpKernel<paddle::platform::CPUDeviceContext, double>}]::<lambda(const class paddle::framework::ExecutionContext&)>'
[04:00:48][Step 1/1] /paddle/paddle/fluid/framework/op_registry.h:103:38:   required from 'void paddle::framework::OpKernelRegistrarFunctor<PlaceType, false, I, KernelTypes ...>::operator()(const char*, const char*, int) const [with PlaceType = paddle::platform::CPUPlace; long unsigned int I = 0ul; KernelTypes = {paddle::operators::AdamOpKernel<paddle::platform::CPUDeviceContext, float>, paddle::operators::AdamOpKernel<paddle::platform::CPUDeviceContext, double>}]'
[04:00:48][Step 1/1] /paddle/paddle/fluid/framework/op_registry.h:130:9:   required from 'paddle::framework::OpKernelRegistrar<PlaceType, KernelType>::OpKernelRegistrar(const char*, const char*, int) [with PlaceType = paddle::platform::CPUPlace; KernelType = {paddle::operators::AdamOpKernel<paddle::platform::CPUDeviceContext, float>, paddle::operators::AdamOpKernel<paddle::platform::CPUDeviceContext, double>}]'
[04:00:48][Step 1/1] /paddle/paddle/fluid/operators/optimizers/adam_op.cc:144:637:   required from here
[04:00:48][Step 1/1] [ 26%] Linking CXX static library libadadelta_op.a
[04:00:48][Step 1/1] /paddle/paddle/fluid/operators/optimizers/adam_op.h:316:13: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
[04:00:48][Step 1/1] Scanning dependencies of target rmsprop_op
[04:00:48][Step 1/1]        if (i == *(rows_ + j)) {
[04:00:48][Step 1/1]              ^
[04:00:48][Step 1/1] /paddle/paddle/fluid/operators/optimizers/adam_op.h:317:31: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
[04:00:48][Step 1/1] [ 26%] Building CXX object paddle/fluid/operators/metrics/CMakeFiles/precision_recall_op.dir/precision_recall_op.cc.o
[04:00:48][Step 1/1]          for (size_t k = 0U; k != row_numel_; ++k) {
[04:00:48][Step 1/1]                                ^
[04:00:48][Step 1/1] /paddle/paddle/fluid/operators/optimizers/adam_op.h:323:31: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
[04:00:48][Step 1/1]          for (size_t k = 0U; k != row_numel_; ++k) {
[04:00:48][Step 1/1] [ 26%] Built target fused_embedding_fc_lstm_op
[04:00:48][Step 1/1]                                ^
[04:00:48][Step 1/1] Scanning dependencies of target proximal_gd_op
[04:00:48][Step 1/1] /paddle/paddle/fluid/operators/optimizers/adam_op.h: In instantiation of 'void paddle::operators::SparseAdamFunctor<T, paddle::operators::CPUAdam>::operator()(size_t) const [with T = double; size_t = long unsigned int]':
[04:00:48][Step 1/1] /paddle/paddle/fluid/operators/optimizers/adam_op.h:487:18:   required from 'void paddle::operators::AdamOpKernel<DeviceContext, T>::Compute(const paddle::framework::ExecutionContext&) const [with DeviceContext = paddle::platform::CPUDeviceContext; T = double]'
[04:00:48][Step 1/1] /paddle/paddle/fluid/framework/op_registry.h:107:11:   required from 'paddle::framework::OpKernelRegistrarFunctor<PlaceType, false, I, KernelTypes ...>::operator()(const char*, const char*, int) const::<lambda(const paddle::framework::ExecutionContext&)> [with PlaceType = paddle::platform::CPUPlace; long unsigned int I = 1ul; KernelTypes = {paddle::operators::AdamOpKernel<paddle::platform::CPUDeviceContext, float>, paddle::operators::AdamOpKernel<paddle::platform::CPUDeviceContext, double>}]'
[04:00:48][Step 1/1] /paddle/paddle/fluid/framework/op_registry.h:106:10:   required from 'struct paddle::framework::OpKernelRegistrarFunctor<PlaceType, false, I, KernelTypes ...>::operator()(const char*, const char*, int) const [with PlaceType = paddle::platform::CPUPlace; long unsigned int I = 1ul; KernelTypes = {paddle::operators::AdamOpKernel<paddle::platform::CPUDeviceContext, float>, paddle::operators::AdamOpKernel<paddle::platform::CPUDeviceContext, double>}]::<lambda(const class paddle::framework::ExecutionContext&)>'
[04:00:48][Step 1/1] /paddle/paddle/fluid/framework/op_registry.h:103:38:   required from 'void paddle::framework::OpKernelRegistrarFunctor<PlaceType, false, I, KernelTypes ...>::operator()(const char*, const char*, int) const [with PlaceType = paddle::platform::CPUPlace; long unsigned int I = 1ul; KernelTypes = {paddle::operators::AdamOpKernel<paddle::platform::CPUDeviceContext, float>, paddle::operators::AdamOpKernel<paddle::platform::CPUDeviceContext, double>}]'
[04:00:48][Step 1/1] /paddle/paddle/fluid/framework/op_registry.h:112:9:   required from 'void paddle::framework::OpKernelRegistrarFunctor<PlaceType, false, I, KernelTypes ...>::operator()(const char*, const char*, int) const [with PlaceType = paddle::platform::CPUPlace; long unsigned int I = 0ul; KernelTypes = {paddle::operators::AdamOpKernel<paddle::platform::CPUDeviceContext, float>, paddle::operators::AdamOpKernel<paddle::platform::CPUDeviceContext, double>}]'
[04:00:48][Step 1/1] /paddle/paddle/fluid/framework/op_registry.h:130:9:   required from 'paddle::framework::OpKernelRegistrar<PlaceType, KernelType>::OpKernelRegistrar(const char*, const char*, int) [with PlaceType = paddle::platform::CPUPlace; KernelType = {paddle::operators::AdamOpKernel<paddle::platform::CPUDeviceContext, float>, paddle::operators::AdamOpKernel<paddle::platform::CPUDeviceContext, double>}]'
[04:00:48][Step 1/1] /paddle/paddle/fluid/operators/optimizers/adam_op.cc:144:637:   required from here
[04:00:48][Step 1/1] /paddle/paddle/fluid/operators/optimizers/adam_op.h:316:13: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
[04:00:48][Step 1/1]        if (i == *(rows_ + j)) {
[04:00:48][Step 1/1]              ^
[04:00:48][Step 1/1] /paddle/paddle/fluid/operators/optimizers/adam_op.h:317:31: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
[04:00:48][Step 1/1]          for (size_t k = 0U; k != row_numel_; ++k) {
[04:00:48][Step 1/1]                                ^
[04:00:48][Step 1/1] /paddle/paddle/fluid/operators/optimizers/adam_op.h:323:31: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
[04:00:48][Step 1/1]          for (size_t k = 0U; k != row_numel_; ++k) {
[04:00:48][Step 1/1]                                ^
```
http://ci.paddlepaddle.org/viewLog.html?tab=buildLog&buildTypeId=Paddle_PrCi&buildId=47932&_focus=2192#_state=344

**Comment by jacquesqiao** [comment]
@luotao1 in next PR?

**Comment by luotao1** [comment]
> in next PR?

Sure.

File: paddle/fluid/operators/optimizers/adam_op.h (hunk 5 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/optimizers/adam_op.h]: fix done.

File: paddle/fluid/operators/optimizers/adam_op.h (hunk 4 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/optimizers/adam_op.h]: fix done

File: paddle/fluid/framework/operator.cc [hunk 9]
```diff
@@ -29,6 +29,7 @@ DECLARE_bool(benchmark);
 DEFINE_bool(check_nan_inf, false,
             "Checking whether operator produce NAN/INF or not. It will be "
             "extremely slow so please use this flag wisely.");
+DEFINE_int32(inner_op_parallelism, 0, "number of threads for inner op");
```
Comment (panyx0718) [inline paddle/fluid/framework/operator.cc:32]: we want to make it a config instead of flag in the future.

**Review by panyx0718 — APPROVED** [review]

**Pull request merged**